Zipkin dependency job fails for any malformed span

Zipkin dependency job (spark job) fails when length of span is more than 10k bytes. 

1. Malformed span details should be logged. 
2. If this error can be ignored then job should proceed further.

`05:33:23.556 [task-result-getter-0] WARN  o.a.spark.scheduler.TaskSetManager - Lost task 0.0 in stage 0.0 (TID 0, localhost): java.lang.IllegalArgumentException: Malformed reading Span from TBinary:
        at zipkin.internal.ThriftCodec.exceptionReading(ThriftCodec.java:507)
        at zipkin.internal.ThriftCodec.read(ThriftCodec.java:444)
        at zipkin.internal.ThriftCodec.readSpan(ThriftCodec.java:69)
        at zipkin.dependencies.cassandra.CassandraDependenciesJob.toLinks(CassandraDependenciesJob.java:146)
        at zipkin.dependencies.cassandra.CassandraDependenciesJob.lambda$run$ef192ee8$1(CassandraDependenciesJob.java:131)
        at zipkin.dependencies.cassandra.CassandraDependenciesJob$$Lambda$4.call(Unknown Source)
        at org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$1$1.apply(JavaRDDLike.scala:129)
        at org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$1$1.apply(JavaRDDLike.scala:129)
        at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:396)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at org.apache.spark.util.collection.ExternalSorter.insertAll(ExternalSorter.scala:189)
        at org.apache.spark.shuffle.sort.SortShuffleWriter.write(SortShuffleWriter.scala:64)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:73)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
        at org.apache.spark.scheduler.Task.run(Task.scala:89)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:227)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: 13610 > 10000: possibly malformed thrift
        at zipkin.internal.ThriftCodec.guardLength(ThriftCodec.java:602)
        at zipkin.internal.ThriftCodec.readList(ThriftCodec.java:461)
        at zipkin.internal.ThriftCodec$ListAdapter.read(ThriftCodec.java:488)
        at zipkin.internal.ThriftCodec$ListAdapter.read(ThriftCodec.java:479)
        at zipkin.internal.ThriftCodec$4.read(ThriftCodec.java:291)
        at zipkin.internal.ThriftCodec$4.read(ThriftCodec.java:261)
        at zipkin.internal.ThriftCodec.read(ThriftCodec.java:442)
        ... 17 more`

--- a/cassandra/src/main/java/zipkin/dependencies/cassandra/CassandraDependenciesJob.java
+++ b/cassandra/src/main/java/zipkin/dependencies/cassandra/CassandraDependenciesJob.java
@@ -33,6 +33,8 @@ import java.util.TimeZone;
 import java.util.concurrent.TimeUnit;
 import org.apache.spark.SparkConf;
 import org.apache.spark.SparkContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import scala.Tuple2;
 import scala.runtime.AbstractFunction1;
 import zipkin.Codec;
@@ -48,6 +50,7 @@ import static zipkin.internal.Util.checkNotNull;
 import static zipkin.internal.Util.midnightUTC;
 
 public final class CassandraDependenciesJob {
+  static final Logger logger = LoggerFactory.getLogger(CassandraDependenciesJob.class);
 
   public static Builder builder() {
     return new Builder();
@@ -143,7 +146,13 @@ public final class CassandraDependenciesJob {
   static Iterable<DependencyLink> toLinks(long startTs, long endTs, Iterable<CassandraRow> rows) {
     List<Span> spans = new LinkedList<>();
     for (CassandraRow row : rows) {
-      spans.add(Codec.THRIFT.readSpan(row.getBytes("span")));
+      try {
+        spans.add(Codec.THRIFT.readSpan(row.getBytes("span")));
+      } catch (RuntimeException e) {
+        logger.warn(String.format(
+            "Unable to decode span from traces where trace_id=%s and ts=%s and span_name='%s'",
+            row.getLong("trace_id"), row.getDate("ts").getTime(), row.getString("span_name")), e);
+      }
     }
     spans = MergeById.apply(spans); // merges the spans by id and also tries to set a timestamp
 

--- a/elasticsearch/src/test/java/zipkin/storage/elasticsearch/ElasticsearchDependenciesTest.java
+++ b/elasticsearch/src/test/java/zipkin/storage/elasticsearch/ElasticsearchDependenciesTest.java
@@ -14,6 +14,7 @@
 package zipkin.storage.elasticsearch;
 
 import com.google.common.util.concurrent.Futures;
+import java.io.IOException;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Set;
@@ -36,7 +37,7 @@ public class ElasticsearchDependenciesTest extends DependenciesTest {
     return storage;
   }
 
-  @Override public void clear() {
+  @Override public void clear() throws IOException {
     storage.clear();
   }
 

--- a/elasticsearch/src/test/java/zipkin/storage/elasticsearch/ElasticsearchTestGraph.java
+++ b/elasticsearch/src/test/java/zipkin/storage/elasticsearch/ElasticsearchTestGraph.java
@@ -22,16 +22,14 @@ enum ElasticsearchTestGraph {
 
   final String index = "test_zipkin";
 
-  static {
-    ElasticsearchStorage.FLUSH_ON_WRITES = true;
-  }
-
   final LazyCloseable<ElasticsearchStorage> storage = new LazyCloseable<ElasticsearchStorage>() {
     public AssumptionViolatedException ex;
 
     @Override protected ElasticsearchStorage compute() {
       if (ex != null) throw ex;
-      ElasticsearchStorage result = new ElasticsearchStorage.Builder().index(index).build();
+      ElasticsearchStorage result = ElasticsearchStorage.builder()
+          .flushOnWrites(true)
+          .index(index).build();
       CheckResult check = result.check();
       if (check.ok) return result;
       throw ex = new AssumptionViolatedException(check.exception.getMessage());

--- a/pom.xml
+++ b/pom.xml
@@ -37,7 +37,7 @@
 
     <scala.binary.version>2.11</scala.binary.version>
     <spark.version>1.6.2</spark.version>
-    <zipkin.version>1.11.0</zipkin.version>
+    <zipkin.version>1.12.1</zipkin.version>
     <logback.version>1.1.7</logback.version>
     <junit.version>4.12</junit.version>
     <assertj.version>3.5.1</assertj.version>

